### ISSUE TITLE
context/tracing in the blockstore/datastore pipeline

### ISSUE THREAD

**Issue by MichaelMure** [ipfs/go-ipfs#6803]
At Infura, we have to deal with the occasional performance issue. Even though we now have added tracing in the HTTP handler of `go-ipfs-cmds` (not upstreamed yet, it's opentracing and you are aiming for opencensus if I'm not mistaken?), the tracing essentially stop there.

The majority of the requests (and the perf issues) involve the data pipeline but it is essentially a black box. To resolve that problem, a proper tracing instrumentation there would be very helpful, but that imply adding a go context to most if not all blockstore and datastore functions.

Is that something you would be interested to pursue ?

cc @dirkmc maybe ?

**Comment by Stebalien** [comment]
We should be passing contexts through to the datastores but the refactor is going to be quite large and invasive and will affect multiple projects (ipfs, libp2p, filecoin, and probably quite a few more).

We should look into this next year but I don't have time right now to handle all the fallout of this refactor.

**Comment by MichaelMure** [comment]
master issue:

Datastore:
- go-datastore: https://github.com/ipfs/go-datastore/pull/161
- go-ds-flatfs: https://github.com/ipfs/go-ds-flatfs/pull/85
- go-ds-badger: https://github.com/ipfs/go-ds-badger/pull/97
- go-ds-leveldb: https://github.com/ipfs/go-ds-leveldb/pull/45
- go-ds-s3:
- go-ds-crdt:
- go-ds-sql:
- go-ds-measure: https://github.com/ipfs/go-ds-measure/pull/30
- go-ds-redis:
- go-ds-swift:
- go-ds-bolt:
- go-ds-badger2:
- go-ds-bitcask:
- go-ds-bench: 

Blockstore:
- go-ipfs-blockstore: https://github.com/ipfs/go-ipfs-blockstore/pull/55

IPFS:
- go-ipfs: https://github.com/ipfs/go-ipfs/pull/7558
- go-ipfs-provider: https://github.com/ipfs/go-ipfs-provider/pull/27 TODO TEST BROKEN
- go-ipfs-routing: https://github.com/ipfs/go-ipfs-routing/pull/23
- go-blockservice: https://github.com/ipfs/go-blockservice/pull/66
- go-bitswap: https://github.com/ipfs/go-bitswap/pull/421
- go-filestore: https://github.com/ipfs/go-filestore/pull/36
- go-graphsync: https://github.com/ipfs/go-graphsync/pull/78
- go-ipfs-exchange-offline: https://github.com/ipfs/go-ipfs-exchange-offline/pull/30
- go-ipfs-exchange-interface: https://github.com/ipfs/go-ipfs-exchange-interface/pull/11
- go-merkledag: https://github.com/ipfs/go-merkledag/pull/57
- go-ipfs-pinner: https://github.com/MichaelMure/go-ipfs-pinner/pull/2

libp2p:
- go-libp2p-kad-dht: https://github.com/libp2p/go-libp2p-kad-dht/pull/683
- go-libp2p-pubsub-router: https://github.com/libp2p/go-libp2p-pubsub-router/pull/78

**Comment by MichaelMure** [comment]
Overall it went fairly well with just go-graphsync to be worked around with `context.TODO().

Obviously this will need to be merged progressively, tagged and updated on the upper layers, but it does compile for me and seems to be working from a quick test.

**Comment by BigLep** [comment]
@MichaelMure : We're sorry this has been open for so long.  Do you think you'd be available to work with a PL engineer later in May get this updated and pushed?  We'll figure out the exact schedule if you're available, but wanted to see if that is possible.  Thanks!

**Comment by MichaelMure** [comment]
@BigLep I might be but I'll have to see how things evolve on my side.

However I'm not sure I would be so useful: the problem here is not technical. Ignoring go-graphsync for a minute, coding wise it's actually fairly easy. It took me a few hours to code and open all those PRs. The few packages I disn't touch are very likely to be as easy.

The problem here is the coordination of all the project and their maintainers to land those changes without too much complications and someone to orchestrate that effort. It's not something I can do as an outsider.

**Comment by MichaelMure** [comment]
@BigLep that said, let me reiterate that this would be a big win for us, but not only. We are often facing performance issue or simply something not working right, and go-ipfs being a black box make it hard to understand and fix. Tracing has proved to be invaluable in other part of our stack to provide a quality service and react quickly on incidents. This is likely the feature that would help us the most.

In addition, this would allow us to give you a better feedback from our deployed infrastructure, like pin-pointing precisely a performance issue. It's always easier to design or fix things with hard numbers.

**Comment by BigLep** [comment]
@MichaelMure : thanks!  The need makes sense and agreed that a key dependency is the coordination from a maintainer to land these.  I'm hoping we secure a time later in the month with someone like @aschmahmann who will do the merging but can also =have you on hand if there rebasing/updates that need to be done along the way.  Does that make sense (and please let me know if that's not an effective strategy)?

**Comment by MichaelMure** [comment]
@BigLep end of May is very much here now, is that still on your radar? On my side I'm still quite busy but this is a quite important topic.

**Comment by BigLep** [comment]
@MichaelMure : thanks for checking in.  This is still on PL's radar.  I'd like to do it after we get the [IPFS 0.9 release](https://github.com/ipfs/go-ipfs/issues/8058) out.  This is taking longer than planned.  I think the week of 2021-06-07 is more likely.  Any chance that would work for you?

**Comment by MichaelMure** [comment]
That looks like a reasonable plan, although I have to mention that I can't guarantee it.

**Comment by BigLep** [comment]
2021-07-01 conversation:
1. We are thinking we should do a Go major bump for these packages.  
2. More thought/planning needs to be done around the names/versions for all the packages

Next steps: @aschmahmann create a new issue for creating the name/version plan.

**Comment by BigLep** [comment]
@MichaelMure : my apologies for not responding by end of last week on where we're at with this.  

We discussed this internally some last week. We realized we need to write up a plan for how we plan to name/version these packages.  Giving more thought here to names/versions may enable us to merge in a more progressive way (rather than the stop-the-world until everything is merged approach like we were originally considering).  @aschmahmann is going to create the issue that describes what needs to be done here.  At the minimum, for go-ipfs 0.10 we want to have this plan articulated.  We are giving priority though to improving our ability to release quickly and likely won't have implemented the plan by then.

We'll update here when we have this issue.  We're tracking creating the issue for the plan [here](https://github.com/orgs/ipfs/projects/9#card-64480215)

**Comment by BigLep** [comment]
Update on where this is at.  This won't happen for at least another couple of go-ipfs releases.  It's currently tagged for 0.13, although the scope of that release is still TBD.  We know it's not in 0.11 or 0.12.  The PL team's priorities have been to focus on our release velocity (but automating and simplifying the release process) and operational issues (in part related to the ipfs.io outage discussed [here](https://blog.ipfs.io/2021-08-27-IPFS.io-gateway-outage-resolution/).

**Comment by BigLep** [comment]
@MichaelMure : quick update here that I'm trying to see about prioritizing this issue sooner as PL is focusing on its own efforts to improve the Gateways it's operating.  @guseggert is a newer team member who may assign this to.  Could you two connect and swap notes about the intended ways this will be helpful?  I'll use that info for making the prioritization case.

**Comment by MichaelMure** [comment]
Sure, here is a few things where that could be helpful:

### tracing/observability

This would be a massive help for running IPFS in production. Really, I can't overstate that. At the moment, go-ipfs is basically a black box: requests comes in, response comes out. What happen in the middle is hard to observe. There is prometheus metrics, the `diag` command or `pprof` but that's really surface level. This leads to difficulties to address emergency situations, understand and plan for the future.

The most critical subsystem in go-ipfs for performance and day to day operation is the data pipeline. Having a go context in there would allow to gradually and possibly independently add tracing instrumentation. The benefits would be:
- full visibility into operational activities: timing, request count, load, delay, inter-dependencies ...
- ability to slice and dice data, isolate specific requests or class of requests, figure out patterns ...
- track errors and problems, isolate their origin and resolve issue in a way simpler and efficient manner

Note also that this tracing would not necessarily be limited to go-ipfs: distributed tracing allow to propagate this tracing over the boundaries of connected systems (proxy, backend ...) which gives another dimension of observability.

### handling cancellation / reliability

At the moment, there is no cancellation in the data pipeline. This means that once a request is started, nothing will stop it, even if the original request is gone. Fixing that would allow to trim that unnecessary fat, reduce the load and improve reliability. It might also prevent a form of attack where heavy handling is triggered with minimal effort.

### request tagging / custom handling

Another way this could be helpful is that the go context allow to carry metadata about the request, **independently of all the layers it's going through**. This means that one could for example carry over a domain specific logger or tag a request with an origin or some customer information. When reaching a lower level, those information could be used to prioritize requests, do caching differently, route to a specialized backend ...

Importantly, this mechanism allows to extend the system **independently** without being bound by the Protocol Labs road map. This would lower your burdens and allow to explore more areas.

### engineering feedback / optimisation

In my experience, each time observability improve, new issues are unearthed. Proper instrumentation would give node operator and PL the tooling to discover those long standing performance/reliability issues. 95% of the work in software engineering is to figure out where the problem comes from. Once that's done and understood, fixing stuff becomes easy.

### node operator autonomy

Observability would allow node operator to more easily figure out what is happening and in turn, rely less on PL to diagnose those issues and reduce the burden on the development team. 

Also discussed a bit at https://github.com/ipfs/roadmap/issues/74

**Comment by BigLep** [comment]
Well stated @MichaelMure .  PL will respond back by EOD 2021-10-01.

**Comment by BigLep** [comment]
I missed not circling back on this last week since we did discuss it internally.  @guseggert and team are going to pick this up.  We're aiming to get this in the next go-ipfs release 0.11.  The first step is to make the plan of how we can merge this in a progressive way.

**Comment by guseggert** [comment]
I've finished the majority of the plumbing, the rest are blocked on two things:

* release of go-libp2p version 0.16.0-rc.1  (to plumb through the changes to go-libp2p)
* releasing the sharding work in https://github.com/ipfs/go-ipfs/issues/8106  (the sharding work contains breaking changes that need to be merged with this)

Once those are resolved, I can complete the plumbing and we will be ready to release go-ipfs v0.11.0-RC1

**Comment by guseggert** [comment]
I discovered yesterday that `go-ipfs-blockstore` is using the wrong version of `go-ipfs-ds-help`, it was inadvertently upgraded to v1, so I need to go back and downgrade `go-ipfs-blockstore@v0` to use `go-ipfs-ds-help@v0` and then re-plumb.

**Comment by aschmahmann** [comment]
closed by #8563
